### DISCUSSION TITLE
Rendering LaTeX in mermaid flowchart

### DISCUSSION THREAD

**Discussion by ellafm** [quarto-dev/quarto-cli#10851] [Q&A] [answered]
### Description

````qmd
---
title: "Some title"
format: html
---

```{mermaid}
flowchart LR
  A[$$x^2$$] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```
````

Results in:
![snapshot](https://github.com/user-attachments/assets/2f1ccebc-5b52-4618-914b-4df9285e1c9e)

The latex is not rendered, but this issue is not present when the flowchart is rendered in mermaid live editor, as shown
[![](https://mermaid.ink/img/pako:eNpFz7EKgzAQBuBXCYdDC7p0dChU7dYudjQWgjk1YHIlJtgivnujpe12990Pxz9DQxIhhXagqemFdexScsPYqYqi5_0QRTVLkiPLdiV5IxnKDvfrPds4nwts1KjILCvmGxZViaMfHCOD9Z_PX3YTbQwxaLRaKBn-z6twcD1q5JCGUWIrQpwDN0uICu_o9jINpM56jMGS73pIWzGMYfMPKRwWSnRW6J-iVI7s9dNwK7q8AUrATVQ?type=png)](https://mermaid.live/edit#pako:eNpFz7EKgzAQBuBXCYdDC7p0dChU7dYudjQWgjk1YHIlJtgivnujpe12990Pxz9DQxIhhXagqemFdexScsPYqYqi5_0QRTVLkiPLdiV5IxnKDvfrPds4nwts1KjILCvmGxZViaMfHCOD9Z_PX3YTbQwxaLRaKBn-z6twcD1q5JCGUWIrQpwDN0uICu_o9jINpM56jMGS73pIWzGMYfMPKRwWSnRW6J-iVI7s9dNwK7q8AUrATVQ)

I'm using quarto version 1.5.56, using VS Code on Windows 10

**Comment by ellafm** [comment]
I think I figured this out. According to the release notes, quarto has been upgraded to natively work with mermaid 10.2.0, but this version does not seem to support latex in flow diagrams. I have managed to get this working as intended by including the following script:

```
        include-in-header:
            text: |
                <script src="https://cdn.jsdelivr.net/npm/mermaid@11.2.1/dist/mermaid.min.js"></script>
```

**Reply by ellafm** [reply]
Updating C:\xxxx\Quarto\share\formats\html\mermaid\mermaid.min.js to the latest version seems to do the trick. 
How can I ask for this to be addressed in the next update?

**Reply by cscheid** [reply]
It's already updated on `main`.

**Reply by ellafm** [reply]
Thanks cscheid. I see this was done after August 29, which explains why I didn't get it when I downloaded the current windows version on this page:
https://quarto.org/docs/get-started/

Does that mean it will be in 1.5.58 ?

**Reply by cscheid** [reply]
In 1.6.12.

**Reply by ellafm** [reply]
OK, thanks. In general, would you recommend downloading the pre-release builds then?

**Reply by harrelfe** [reply]
I just installed Quarto 1.6.12 and it says it's using `mermaid` 11.2.0 and I get a syntax error with

````
```{mermaid}
flowchart LR
  A[$$\alpha_{3} + \beta\times X$$] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```
````

**Reply by cscheid** [reply]
I get that error as well. It's possible that this was added from 11.2.0 to 11.2.1? I'm not a mermaidjs expert and their error diagnostics don't give much help, unfortunately.

EDIT: The error remains on 11.2.1. 🤷 It's possible that the way mermaid's live code editor works is that it passes LaTeX along and then calls mathjax on the resulting file. We don't do that right now and I don't expect us to do it in the near term.

**Reply by cscheid** [reply]
edit: I get the same error on mermaid's live editor, so I don't know how @ellafm achieved this.

**Reply by cscheid** [reply]
<img width="997" alt="image" src="https://github.com/user-attachments/assets/a824efbd-7767-4357-9e52-5df58f886a33">

**Reply by mcanouil** [reply]
The issue is the use of the brackets or more accurately, because the expression is unquoted.
This is all illustrated in the documentation, <https://mermaid.js.org/config/math.html>.

Note that as with unquoted YAML string, using unquoted mermaid strings leads to parsing issue.
Mermaid recommends (see the documentation) the use of quotes for anything not ASCII (including markdown formatting and math).

**Reply by cscheid** [reply]
Thank you! That worked:

<img width="651" alt="image" src="https://github.com/user-attachments/assets/0d22596b-4d9a-435b-9882-125d661b13b4">

````

```{mermaid}
flowchart LR
  A["$$\alpha_{3} + \beta\times X$$"] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```
````

**Reply by mcanouil** [reply]
And for everyone that will try "LaTeX" things, as documented, Mermaid.js is using KaTeX, so you also need to understand what KaTeX supports before reporting issues/bugs (https://katex.org/).

**Reply by ellafm** [reply]
Incidentally, I think there may also be a bug in mermaid 11.2.1 with how it interprets escape characters. See:
[https://github.com/orgs/mermaid-js/discussions/5885](https://github.com/orgs/mermaid-js/discussions/5885)

**Reply by felixvuo** [reply]
I'm also struggling to get LaTeX rendered from quarto+mermaid - I am using the pdf output.  Here's my source:
    ---
    title: "Important Concepts"
    format:
      pdf:
        documentclass: scrartcl
    ---
    
    # An Example
    
    ```{mermaid}
    %%| label: fig-tree
    %%| fig-cap: An important graph
    graph TD
        A("$$x+y+z \leq 5$$")
        A -- "$$x=1$$" --> A1
        A -- "$$x=2$$" --> A2
        A -- "$$x=3$$" --> A3
    ```
    
    In @fig-tree we can see a beautiful thing.

I am using quarto 1.5.57, the latest `.deb` I found for Ubuntu.  The rendered output looks like this:
![Screenshot from 2024-09-24 11-20-41](https://github.com/user-attachments/assets/53e8b7fc-64fa-4215-afc5-73bf4c247921)

I have tried with or without the double quotes, same output.  Any further suggestions?  E.g. can I point quarto to the latest mermaid version somehow?

**Reply by cderv** [reply]
@felixvuo from answers just above I think you need pre-release 1.6.12 to get latest mermaid.js

Please do try it: quarto.org/docs/download/prerelease.html

**Reply by felixvuo** [reply]
@cderv thank you, it does seem to work fine with 1.6.12.

**Reply by mcanouil** [reply]
Cannot reproduce your issue.
<img width="1624" alt="image" src="https://github.com/user-attachments/assets/e3863416-b974-469a-825c-50c7b5021bc9">

**Comment by tony-an-phd** [comment]
I am using quarto 1.6.33 to create a flowchat with math inside a box using mermaid with the following code in a file called n.qmd: 
````qmd
---
format:
  html:
    theme: default
  pdf:
    documentclass: article
keep-tex: true
---

```{mermaid}
graph LR
    A("$$x+y+z \leq 5$$")
    A -- "$$x=1$$" --> A1
    A -- "$$x=2$$" --> A2
    A -- "$$x=3$$" --> A3
```
````
Once I issued the command `quarto render n.qmd` I see that n.pdf shows correctly, and the n.tex shows it's showing that png file with math inside the box A:
`\includegraphics[width=3.05in,height=2.9in]{n_files/figure-latex/mermaid-figure-1.png}`

However, the n.html file generated using mermaid.js shows an empty box for A:

````html
<div class="cell" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class="figure"><p></p>
<div>
<pre class="mermaid mermaid-js">graph LR
    A("$$x+y+z \leq 5$$")
    A -- "$$x=1$$" --&gt; A1
    A -- "$$x=2$$" --&gt; A2
    A -- "$$x=3$$" --&gt; A3
</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
````

This the screen shot of n.html in Firefox 132.01:

![386244165-6aa20ede-c143-4c46-9369-a628feaa13c8](https://github.com/user-attachments/assets/8ba2d590-e33e-4d21-b892-0705314c5918)

How do I make the n.html display the math properly?

````txt
> quarto check
Quarto 1.6.33
[>] Checking environment information...
      Quarto cache location: C:\Users\ynots\AppData\Local\quarto
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.4.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.46.3: OK
      Typst version 0.11.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.6.33
      Path: C:\Users\ynots\AppData\Local\Programs\Quarto\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: v2024.11
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Using: TinyTex
      Path: C:\Users\ynots\AppData\Roaming\TinyTeX\bin\windows\
      Version: 2024

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....(None)

      Unable to locate an installed version of Python 3.
      Install Python 3 from https://www.python.org/downloads/

[>] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/
      
````

**Reply by mcanouil** [reply]
Of course HTML is not using the PNG. Mermaid.js is a JavaScript library to produce diagram in HTML.
The PNG is a "screenshot" of that for use in non HTML formats.

Everything is working as expected.

**Reply by mcanouil** [reply]
Could you avoid sharing screenshots with no context?
Where did you take this screenshot? HTML? PDF? Both?

Pasting several times the same thing does not make it clearer.

The code works perfectly at least for HTML as already illustrated:

https://github.com/quarto-dev/quarto-cli/discussions/10851#discussioncomment-11252396

**Reply by mcanouil** [reply]
Where in <https://github.com/quarto-dev/quarto-cli/discussions/10851#discussioncomment-11255530> ? You have already duplicated the discussion in two places. Try to keep your replies in the same place/thread and try to keep context. That's not only for me, that's for anyone that will read this.

We are going in circles here.
You need to provide more context, for example the output of `quarto check`, what's your IDE, how do you view/preview the HTML file, etc.
I cannot reproduce your issue as showed.
So sharing again your screenshot does not help us help you.

**Reply by tony-an-phd** [reply]
I edited the previous post to get everything in one place

**Reply by mcanouil** [reply]
Thanks!

How do you preview the HTML?

**Reply by tony-an-phd** [reply]
I use Firefox 132.01. - added it to the post.

**Reply by cscheid** [reply]
I suspect that you need to run a local server instead of (eg) double-clicking your file on Explorer. That's not a Quarto bug; that's a requirement of javascript libraries that use ES6 modules.

**Reply by tony-an-phd** [reply, marked as answer]
Thank you for the tip @cscheid ! It looks like that it's not working in Firefox! I opened a Chrome and it did displayed correctly.
